prov/rxm/src/rxm_ep.c

```diff
@@ -1056,84 +1013,134 @@ rxm_ep_sar_tx_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn, void *conte
 		}
 	}
+
+	if (dlist_empty(&rxm_conn->deferred_op_list))
+		dlist_remove_init(&rxm_conn->conn_deferred_entry);
 }
 
-static inline ssize_t
-rxm_ep_postpone_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
-		     void *context, uint8_t count, const struct iovec *iov,
-		     void **desc, size_t len, uint64_t data, uint64_t flags,
-		     uint64_t tag, uint8_t op, uint64_t comp_flags)
+void rxm_ep_progress_deferred_list(struct rxm_ep *rxm_ep)
 {
-	struct rxm_tx_entry *tx_entry;
-	struct rxm_tx_buf *tx_buf;
+	struct rxm_conn *rxm_conn;
+	struct dlist_entry *tmp;
+
+	dlist_foreach_container_safe(&rxm_ep->conn_deferred_list,
+				     struct rxm_conn, rxm_conn,
+				     conn_deferred_entry, tmp) {
+		if (OFI_UNLIKELY((rxm_conn->handle.state != CMAP_CONNECTED) &&
+				 (rxm_conn->handle.state != CMAP_CONNECTED_NOTIFY)))
+			continue;
+		rxm_ep_progress_conn_deferred_list(rxm_ep, rxm_conn);
+	}
+}
 
-	FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
-	       "Buffer TX request (len - %zd) for %p conn\n", len, rxm_conn);
+static ssize_t rxm_ep_defer_tx_inject_iov(struct rxm_ep *rxm_ep, const struct iovec *iov,
+					  size_t count, fi_addr_t dest_addr, void *context,
+					  uint64_t data, uint64_t flags, uint64_t tag,
+					  uint8_t op, uint64_t comp_flags, struct rxm_conn *rxm_conn)
+{
+	int ret;
+	size_t len = ofi_total_iov_len(iov, count);
+	struct rxm_tx_entry *tx_entry = calloc(1, sizeof(*tx_entry));
+	if (OFI_UNLIKELY(!tx_entry))
+		return -FI_EAGAIN;
 
-	if (len > rxm_ep->rxm_info->tx_attr->inject_size) {
-		if (rxm_ep_alloc_lmt_tx_res(rxm_ep, rxm_conn, context, count,
-					    iov, desc, len, data, flags,
-					    comp_flags, tag, op, &tx_entry) < 0)
-			return -FI_EAGAIN;
-	} else {
-		ssize_t ret = rxm_ep_format_tx_res(rxm_ep, rxm_conn, context, count,
-						   len, data, flags, comp_flags,
-						   tag, &tx_buf, &tx_entry,
-						   &rxm_ep->buf_pools[RXM_BUF_POOL_TX]);
-		if (OFI_UNLIKELY(ret))
-			return ret;
-		tx_buf->pkt.hdr.op = op;
-		tx_buf->pkt.hdr.flags = comp_flags;
-		ofi_copy_from_iov(tx_buf->pkt.data, tx_buf->pkt.hdr.size,
-				  iov, count, 0);
-		tx_entry->state = RXM_TX;
+	ret = rxm_ep_format_tx_res_lightweight(rxm_ep, rxm_conn, len, data,
+					       flags, tag, &tx_entry->tx_buf,
+					       ((len + sizeof(struct rxm_pkt) <=
+						rxm_ep->msg_info->tx_attr->inject_size) ?
+						&rxm_ep->buf_pools[RXM_BUF_POOL_TX_INJECT] :
+						&rxm_ep->buf_pools[RXM_BUF_POOL_TX]));
+	if (ret)
+		goto err1;
+	tx_entry->state = RXM_TX;
+	rxm_fill_tx_entry(context, count, flags, comp_flags, tx_entry->tx_buf, tx_entry);
+	tx_entry->tx_buf->pkt.hdr.op = op;
+	tx_entry->tx_buf->pkt.hdr.flags = comp_flags;
+
+	ofi_copy_from_iov(tx_entry->tx_buf->pkt.data, tx_entry->tx_buf->pkt.hdr.size,
+			  iov, count, 0);
+
+	rxm_ep->util_ep.tx_cq->cq_fastlock_acquire(&rxm_ep->util_ep.tx_cq->cq_lock);
+	dlist_insert_tail(&tx_entry->deferred_entry, &rxm_conn->deferred_op_list);
+	if (dlist_empty(&rxm_conn->conn_deferred_entry)) {
```

Comment:
I'm using `rxm->conn_deffered_entry` in two ways:
- as a flag to check whether this entry is in `rxm_ep->conn_deferred_list` -- this helps to avoid going around the list and introducing new variable.
- as an entry to be inserted into `rxm_ep->conn_deferred_list` that notifies CQ polling to progress deferred queue of this connection.


p.s. your comment helps me to understand that this is not clear solution and should be re-implemented. Thank you!
Now the solution is to check whether the `rxm_conn->deferred_op_list` is empty:
- if this empty, we insert `rxm_conn->conn_deferred_entry` to the `rxm_ep->conn_deferred_list` to notify CQ handling function
- if this isn't empty, nothing should be done.